File: Python/10-BitManipulation/Algorithms/1-Bitwise-AND-OR-XOR/Bitwise-AND-OR-XORPro.py
def print_binary(number):
    """
    Helper function to print the binary representation of a number, padded to 8 bits.

    Args:
        number (int): The integer to print in binary.
    """
    if not isinstance(number, int):
        print("Invalid input: Please provide an integer.")
        return
    if number < 0:
        print("Negative numbers are not handled in this simplified binary representation.")
        return
    print(f"{number:08b}")  # Pad with zeros to 8 bits.  For larger numbers, adjust the 8.

def bitwise_and_example(num1, num2):
    """
    Example of Bitwise AND (&) operation.

    Args:
        num1 (int): The first integer.
        num2 (int): The second integer.

    Returns:
        int: The result of the bitwise AND operation.
    """
    print("\nBitwise AND Example:")
    print(f"Number 1: {num1} (Binary: ", end="")
    print_binary(num1)
    print(f"Number 2: {num2} (Binary: ", end="")
    print_binary(num2)
    result = num1 & num2
    print(f"Result: {result} (Binary: ", end="")
    print_binary(result)
    return result

def bitwise_or_example(num1, num2):
    """
    Example of Bitwise OR (|) operation.

    Args:
        num1 (int): The first integer.
        num2 (int): The second integer.

    Returns:
        int: The result of the bitwise OR operation.
    """
    print("\nBitwise OR Example:")
    print(f"Number 1: {num1} (Binary: ", end="")
    print_binary(num1)
    print(f"Number 2: {num2} (Binary: ", end="")
    print_binary(num2)
    result = num1 | num2
    print(f"Result: {result} (Binary: ", end="")
    print_binary(result)
    return result

def bitwise_xor_example(num1, num2):
    """
    Example of Bitwise XOR (^) operation.

    Args:
        num1 (int): The first integer.
        num2 (int): The second integer.

    Returns:
        int: The result of the bitwise XOR operation.
    """
    print("\nBitwise XOR Example:")
    print(f"Number 1: {num1} (Binary: ", end="")
    print_binary(num1)
    print(f"Number 2: {num2} (Binary: ", end="")
    print_binary(num2)
    result = num1 ^ num2
    print(f"Result: {result} (Binary: ", end="")
    print_binary(result)
    return result

def bit_masking_example():
    """
    Example of Bit Masking to extract specific bits.
    Real-world application: Reading status flags from a register in embedded systems.
    """
    print("\nBit Masking Example:")
    status_register = 0b11010110  # Example status register value
    print(f"Status Register: {status_register} (Binary: ", end="")
    print_binary(status_register)

    # Define masks for the bits we want to extract
    mask_bit_3 = 0b00001000  # Mask for bit 3 (remember, we start at bit 0)
    mask_bit_5 = 0b00100000  # Mask for bit 5
    mask_bits_0_2 = 0b00000111 # Mask for bits 0, 1, and 2

    # Extract the bits using bitwise AND
    bit_3_value = (status_register & mask_bit_3) >> 3  # Right shift to get the bit value
    bit_5_value = (status_register & mask_bit_5) >> 5
    bits_0_2_value = status_register & mask_bits_0_2

    print(f"Bit 3 Value: {bit_3_value}")
    print(f"Bit 5 Value: {bit_5_value}")
    print(f"Bits 0-2 Value: {bits_0_2_value}")

def bit_toggling_example():
    """
    Example of Bit Toggling using XOR.
    Real-world application: Flipping the state of a GPIO pin in embedded systems,
                       or toggling a boolean flag in a program.
    """
    print("\nBit Toggling Example:")
    control_register = 0b10101010  # Example control register
    print(f"Control Register: {control_register} (Binary: ", end="")
    print_binary(control_register)

    # Toggle bit 2 (index 2)
    toggle_mask = 0b00000100
    control_register = control_register ^ toggle_mask
    print(f"After toggling bit 2: {control_register} (Binary: ", end="")
    print_binary(control_register)

    # Toggle bit 5 (index 5)
    toggle_mask = 0b00100000
    control_register = control_register ^ toggle_mask
    print(f"After toggling bit 5: {control_register} (Binary: ", end="")
    print_binary(control_register)
    return control_register

def check_if_power_of_two(n):
    """
    Checks if a number is a power of two using bitwise AND.
    Real-world application: Optimizing certain algorithms or data structures
                         that rely on powers of two (e.g., memory allocation,
                         fast Fourier transform).
    """
    print("\nCheck if Power of Two:")
    print(f"Number: {n}")
    if n <= 0:
        print(f"{n} is not a power of two.")
        return False
    # A power of two has only one bit set.  n & (n-1) clears the least significant bit.
    if (n & (n - 1)) == 0:
        print(f"{n} is a power of two.")
        return True
    else:
        print(f"{n} is not a power of two.")
        return False

def count_set_bits(n):
    """
    Counts the number of set bits (1s) in an integer using bitwise operations.
    Brian Kernighan's Algorithm
    Real-world application: Determining the Hamming weight of a number,
                         which has applications in cryptography, error detection,
                         and image processing.
    """
    print("\nCount Set Bits:")
    print(f"Number: {n} (Binary: ", end="")
    print_binary(n)
    count = 0
    while n > 0:
        n &= (n - 1)  # Clear the least significant set bit
        count += 1
    print(f"Number of set bits: {count}")
    return count

def main():
    """
    Main function to execute the examples.
    """
    num1 = 10  # Binary: 1010
    num2 = 6   # Binary: 0110

    bitwise_and_example(num1, num2)
    bitwise_or_example(num1, num2)
    bitwise_xor_example(num1, num2)
    bit_masking_example()
    bit_toggling_example()
    check_if_power_of_two(16)
    check_if_power_of_two(20)
    count_set_bits(0b10110101)  # Binary: 10110101 (7 set bits)

if __name__ == "__main__":
    main()

File: Java/10-BitManipulation/Algorithms/1-Bitwise-AND-OR-XOR/Bitwise-AND-OR-XORPro.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class BitwiseOperations {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int num1, num2;

        System.out.println("Enter two integers:");
        num1 = scanner.nextInt();
        num2 = scanner.nextInt();

        System.out.println("\nBitwise Operations:");
        System.out.println("-------------------");

        // 1. Basic Bitwise AND, OR, XOR
        System.out.println("\n1. Basic Operations:");
        basicOperations(num1, num2);

        // 2. Using Bitwise AND for Checking Even/Odd
        System.out.println("\n2. Checking Even/Odd:");
        checkEvenOdd(num1);

        // 3. Using Bitwise OR for Setting a Bit
        System.out.println("\n3. Setting a Bit:");
        int bitToSet = 2; // Set the 2nd bit (0-based index)
        int resultOr = setBit(num1, bitToSet);
        System.out.println("Original number: " + num1 + ", After setting bit " + bitToSet + ": " + resultOr);

        // 4. Using Bitwise XOR for Toggling a Bit
        System.out.println("\n4. Toggling a Bit:");
        int bitToToggle = 1; // Toggle the 1st bit (0-based index)
        int resultXor = toggleBit(num1, bitToToggle);
        System.out.println("Original number: " + num1 + ", After toggling bit " + bitToToggle + ": " + resultXor);

        // 5. Using Bitwise AND to check if a number is power of 2
        System.out.println("\n5. Check if Power of 2:");
        checkPowerOfTwo(num1);

        scanner.close();
    }

    // 1. Basic Bitwise AND, OR, XOR Operations
    // Demonstrates the basic usage of bitwise operators.
    public static void basicOperations(int num1, int num2) {
        System.out.println("Number 1: " + num1 + " (" + Integer.toBinaryString(num1) + ")");
        System.out.println("Number 2: " + num2 + " (" + Integer.toBinaryString(num2) + ")");
        System.out.println("-------------------");
        System.out.println("Bitwise AND (&): " + (num1 & num2) + " (" + Integer.toBinaryString(num1 & num2) + ")");
        System.out.println("Bitwise OR  (|): " + (num1 | num2) + " (" + Integer.toBinaryString(num1 | num2) + ")");
        System.out.println("Bitwise XOR (^): " + (num1 ^ num2) + " (" + Integer.toBinaryString(num1 ^ num2) + ")");
    }

    // 2. Using Bitwise AND for Checking Even/Odd
    // A common application of bitwise AND is to efficiently check if a number is even or odd.
    // If the least significant bit is 0, the number is even; otherwise, it's odd.
    public static void checkEvenOdd(int num) {
        if ((num & 1) == 0) {
            System.out.println(num + " is even.");
        } else {
            System.out.println(num + " is odd.");
        }
    }

    // 3. Using Bitwise OR for Setting a Bit
    // Bitwise OR can be used to set a specific bit in a number to 1.
    public static int setBit(int num, int bitPosition) {
        // Create a mask with a 1 at the desired bit position.
        int mask = 1 << bitPosition;
        return num | mask; // ORing with the mask sets the bit.
    }


    // 4. Using Bitwise XOR for Toggling a Bit
    // Bitwise XOR can be used to toggle a specific bit in a number (change 0 to 1 and vice-versa).
    public static int toggleBit(int num, int bitPosition) {
        // Create a mask with a 1 at the desired bit position.
        int mask = 1 << bitPosition;
        return num ^ mask; // XORing with the mask toggles the bit.
    }

    // 5. Using Bitwise AND to check if a number is power of 2
    public static void checkPowerOfTwo(int num) {
        if (num <= 0) {
            System.out.println(num + " is not a power of 2.");
            return;
        }
        // A power of 2 has only one bit set to 1.  Subtracting 1 will make that bit 0, and all lower bits 1.
        // ANDing the number with its predecessor will result in 0 if it's a power of 2.
        if ((num & (num - 1)) == 0) {
            System.out.println(num + " is a power of 2.");
        } else {
            System.out.println(num + " is not a power of 2.");
        }
    }
}

/*
Explanation:

1. Basic Bitwise Operations:
   - Demonstrates the fundamental use of &, |, and ^.
   - Example:
     - num1 = 5 (0101 in binary)
     - num2 = 3 (0011 in binary)
     - AND (&):  0101 & 0011 = 0001 (1 in decimal)
     - OR  (|):  0101 | 0011 = 0111 (7 in decimal)
     - XOR (^):  0101 ^ 0011 = 0110 (6 in decimal)

2. Checking Even/Odd:
   - Uses (num & 1) to check the least significant bit.
   - If the LSB is 0, the number is even.
   - If the LSB is 1, the number is odd.
   - This is more efficient than using the modulo operator (%)

3. Setting a Bit:
   - Uses the bitwise OR operator (|) to set a specific bit to 1.
   - A mask is created with a 1 at the desired position.
   - ORing the number with the mask sets that bit without affecting others.

4. Toggling a Bit:
   - Uses the bitwise XOR operator (^) to toggle a specific bit.
   - A mask is created with a 1 at the desired position.
   - XORing with the mask flips the bit (0 becomes 1, 1 becomes 0).

5. Check if Power of 2:
   - Uses the property that powers of 2 have only one bit set to 1.
   - num & (num - 1) clears the least significant 1-bit.
   - If the result is 0, the number is a power of 2.
   - Handles the edge case where num is 0.

Real-World Applications:

- Embedded Systems: Bitwise operations are crucial for low-level hardware manipulation, setting registers, and controlling devices.
- Network Programming: Used for manipulating IP addresses, subnet masks, and protocol flags.
- Graphics Programming:  Working with pixel colors, masking, and other graphical operations.
- Cryptography:  Used in encryption algorithms and hash functions.
- Data Compression:  Bitwise operations can be used to efficiently pack and unpack data.
- Game Development:  Used for collision detection, setting game states, and optimizing performance.
*/
